### ISSUE TITLE
Internal error when constructing function type that returns a tuple with size greater than 2

### ISSUE THREAD

**Issue by Guillaume-Helbecque** [chapel-lang/chapel#23168]
### Summary of Problem

I got the following internal error when I tried to construct a function type that returns a tuple with a size greater than 2:
```
$CHPL_HOME/modules/internal/ChapelTuple.chpl:26: internal error: RES-VIS-ONS-1348 chpl version 1.31.0

Internal errors indicate a bug in the Chapel compiler,
and we're sorry for the hassle.  We would appreciate your reporting this bug --
please see https://chapel-lang.org/bugs.html for instructions.  In the meantime,
the filename + line number above may be useful in working around the issue.
```
I was surprised that using `2*int` works (see `Tfoo` below). Is it a known issue?

### Steps to Reproduce

**Source Code:**

```chapel
proc foo(x: int, y: int): 2*int { return (x, y); }
type Tfoo = proc(x: int, y: int): 2*int;
var a: Tfoo = foo;
writeln(a.type:string); // proc(x: int, y: int): _tuple_2_star_int64_t

proc bar(x: int, y: int, z: int): 3*int { return (x, y, z); }
type Tbar = proc(x: int, y: int, z: int): 3*int; // breaks at compile-time
/* var b: Tbar = bar;
writeln(b.type:string); */
```

**Compile command:**
`chpl test.chpl`

**Associated Future Test(s):**

### Configuration Information

- Output of `chpl --version`: `chpl version 1.31.0`
- Output of `$CHPL_HOME/util/printchplenv --anonymize`:
```
CHPL_TARGET_PLATFORM: linux64
CHPL_TARGET_COMPILER: gnu
CHPL_TARGET_ARCH: x86_64
CHPL_TARGET_CPU: native
CHPL_LOCALE_MODEL: flat
CHPL_COMM: none
CHPL_TASKS: qthreads
CHPL_LAUNCHER: none
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: jemalloc
CHPL_ATOMICS: cstdlib
CHPL_GMP: bundled
CHPL_HWLOC: bundled
CHPL_RE2: bundled
CHPL_LLVM: none *
CHPL_AUX_FILESYS: none
```
- Back-end compiler and version, e.g. `gcc --version` or `clang --version`: `gcc (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0`

**Comment by bradcray** [comment]
@Guillaume-Helbecque : Thanks for making us aware of this.  The first-class procedures support ("procedure/function types") is still a new enough feature that it's not surprising that there are bugs in it, and I believe this one was previously unknown to us.  It is odd/surprising that there's this lack of orthogonality, though, between 2D and 3D.

I'm scrambling to get ready for some meetings at the moment, but will try to see whether there's an easy explanation for this later on if nobody else beats me to it first.

One thing I'm finding is that [if I remove the explicit type alias](https://ato.pxeger.com/run?1=dVBLCsJADMXtnCLLVqxCFZEp3sKFu2FaRh3sZ0hTbRVP4saNHsrTOK1W8NNNAkney3vvfIs20qj4crkWtPJm997SYBbBKsuckoNOaQBV010Oft92OAIqKjAFp7Q7N4ATo8ooWFgMzKGGd0ADtpMIkrentgZsj5pUnDpyWLPwnFCna8s6GnVQCSpMrIQvcpIo7Gw6EcRYIzuU-AUYwKEFjv_It-vGQS0s5P8-_hDM6y-NvhCV3OYgCaIsMTpWHulEvR2Fn46eAb9ybvN-AA), I don't get the internal error.  I don't know whether that is an option in the real code that is motivating this small reproducer, though:

```chapel
proc foo(x: int, y: int): 2*int { return (x, y); }
type Tfoo = proc(x: int, y: int): 2*int;
var a: Tfoo = foo;
writeln(a.type:string); // proc(x: int, y: int): _tuple_2_star_int64_t

proc bar(x: int, y: int, z: int): 3*int { return (x, y, z); }
var b: proc(x: int, y: int, z: int): 3*int = bar; // breaks at compile-time
writeln(b.type:string);
```

**Comment by bradcray** [comment]
I did not manage to get back to this last week, but took a quick look at it today.  The following information is probably only useful to developers:
* failure is an "Expression has no visibility block." in resolution code that @mppf developed (though this seems likely to be a case of GIGO)
* the call in question is a call to chpl__autoCopy() that was inserted in resolution as well (`autoMemoryFunction`) for a 3-tuple temp variable
* compiling with `--verify` doesn't trip any new errors
* checking the log of the file just before resolution runs, the code for [T]foo() vs. [T]bar doesn't show any significant differences
* taking out the foo-related 2-tuple code doesn't help (i.e., it doesn't seem to be a case of "doing something twice breaks things")

@Guillaume-Helbecque : Would you let me know whether this issue is blocking you?

**Comment by Guillaume-Helbecque** [comment]
Thank you for the information. I think I can workaround this.

**Comment by DanilaFe** [comment]
@Guillaume-Helbecque : I've merged a PR (#23295) that should address this bug; I'm thus closing the issue. Please feel free to re-open if you believe something is wrong.

**Comment by bradcray** [comment]
Thanks for jumping on this bug Daniel!  And thanks for reporting it Guillaume!
